Nominating and checking validity for `ConsensusData` containing slashed validators

linked0 opened the pull request:
This is implementation about nominating and checking the validity for `ConsensusData` containing the merkleized `preimage_root` of all the preimages and the UTXO indices of validators that are missing to relay their pre-images.

This PR does not have some requirements as follows but the missing requirements will be added in other PRs soon.
* Check for `ConsensusData` which tries to melt a UTXO which is in the MPV
* Changing getRandomSeed

Part of #1076

linked0 committed:
Add information about slashing into `ConsensusParam`

linked0 committed:
Implement class managing slashing policy

The `SlashPolicy` class manages the slashing policy for the misbehaving
validators that do not reveal pre-images timely. The class has two
responsibilites as follows.
- How many times of missing pre-images it will allow
- How many BOAs it will impose on the misbehaving validators as a penalty

TrustHenry commented:
> The `SlashPolicy` clas manages the slashing policy for the misbehaving

clas -> class
There is a typo in the second commit message.

TrustHenry commented on source/agora/consensus/SlashPolicy.d:
```diff
@@ -158,14 +164,67 @@ public class SlashPolicy
+        scope(failure) assert(0);
         auto preimage = this.enroll_man.getValidatorPreimage(utxo_key);
         auto enrolled = this.enroll_man.getEnrolledHeight(preimage.enroll_key);
+
```
?

linked0 replied:
Fixed it.

TrustHenry commented on source/agora/node/Ledger.d:
```diff
@@ -380,6 +385,7 @@ public class Ledger
     protected void updateUTXOSet (const ref Block block) @safe
     {
         const height = block.header.height;
+
```
You can make a separate commit.

linked0 replied:
deleted the unnecessary line.

TrustHenry commented:
I'll have to look at it more.
I think it's good. It's a good start.

linked0 commented:
> > The `SlashPolicy` clas manages the slashing policy for the misbehaving
> 
> clas -> class
> There is a typo in the second commit message.

modified the message.

linked0 commented:
Need to split the last commit into smaller commits.

AndrejMitrovic commented on source/agora/consensus/SlashPolicy.d:
```diff
@@ -0,0 +1,286 @@
+    public void getMissingValidators (ref uint[] missing_validators,
+        in Height height) @safe nothrow
+    {
+        Hash[] keys;
```
If you plan to use this as a buffer then I think you should add:

```
missing_validators.length = 0;
() @trusted { assumeSafeAppend(missing_validators); }();
```

Otherwise if you don't want it to be a buffer you can use `out uint[]`. But it would mean there is always an allocation. So the first approach might be better.

AndrejMitrovic commented on source/agora/consensus/SlashPolicy.d:
```diff
@@ -0,0 +1,286 @@
+
+    ***************************************************************************/
+
+    private bool checkValidValidator (in Height height, in Hash utxo_key)
```
I would name this something better like `hasRevealedPreimage`. Because the concept of a "valid validator" doesn't really sound informative.

AndrejMitrovic commented on source/agora/consensus/SlashPolicy.d:
```diff
@@ -0,0 +1,286 @@
+                valid_keys ~= key;
+        }
+
+        return this.enroll_man.getRandomSeed(valid_keys, height);
```
That's not a merkle root though.

If you want we can just rename it to something like `random_seed` in the `ConsensusData` and in the `Block`. I think it's fine to just store the value of `getRandomSeed()`, it's just that the naming is wrong because that function doesn't actually create a merkle root.

AndrejMitrovic commented on source/agora/consensus/SlashPolicy.d:
```diff
@@ -0,0 +1,286 @@
+
+    /***************************************************************************
+
+        Check if the the validator is valid for the height
+
+        This checks if a validator has a pre-image for the height
```
You can remove the first sentence, the second one already describes what the function does.

AndrejMitrovic commented on source/agora/consensus/SlashPolicy.d:
```diff
@@ -0,0 +1,286 @@
+            utxo_key = the UTXO key idendifying a validator
+
+        Returns:
+            true if the validator has a right pre-image
```
> has a right pre-image

has revealed its preimage for the provided block height

AndrejMitrovic commented on source/agora/consensus/SlashPolicy.d:
```diff
@@ -40,7 +40,8 @@ import agora.consensus.state.UTXODB;
 import agora.utils.Log;
```
The comment in the commit should be changed, because the preimage root is not merkleized. It's just a hash concatenation of all the preimages which were sorted by the utxo key.

AndrejMitrovic commented on source/agora/consensus/protocol/Nominator.d:
```diff
@@ -544,6 +553,21 @@ extern(D):
+            }
+            else
+            {
+                temp_level = ValidationLevel.kMaybeValidValue;
```
`return ValidationLevel.kMaybeValidValue`

AndrejMitrovic commented on source/agora/consensus/protocol/Nominator.d:
```diff
@@ -534,6 +542,7 @@ extern(D):
     {
         const cur_time = this.clock.networkTime();
         const exp_time = this.getExpectedBlockTime(Height(slot_idx));
+        ValidationLevel temp_level;
```
We don't need a state variable for this, we can use early returns. I'll add more comments below.

AndrejMitrovic commented on source/agora/consensus/protocol/Nominator.d:
```diff
@@ -544,6 +553,21 @@ extern(D):
+                    log.error("validateValue(): Preimage Validation failed: {}. Data: {}",
+                        fail_reason, data);
+                    return ValidationLevel.kInvalidValue;
+                }
```
Add `return ValidationLevel. kFullyValidatedValue` here and then you won't need the `temp_level` anymore.

AndrejMitrovic commented on source/agora/node/Ledger.d:
```diff
@@ -494,6 +506,56 @@ public class Ledger
+
+    ***************************************************************************/
+
+    public string validateSlashingData (ConsensusData data) @trusted
```
`in ConsensusData data`

AndrejMitrovic replied:
Can we make this function `@safe` and also mark `isInvalidPreimageRootReason` as `@safe` too if needed? We should avoid labeling everything as `@trusted`, because it tells the compiler to avoid doing any kind of safety checks.

linked0 replied:
> Can we make this function @safe and also mark isInvalidPreimageRootReason as @safe too if needed?

We can. I'm not gonna use `@trust` as far as possible.

AndrejMitrovic commented on source/agora/consensus/protocol/Nominator.d:
```diff
@@ -544,6 +553,21 @@ extern(D):
+            bool is_same = this.ledger.isSameMissingValidatorsAsLocal(data);
+            if (is_same)
+            {
+                if (auto fail_reason = this.ledger.validateSlashingData(data))
```
`isSameMissingValidatorsAsLocal` and `validateSlashingData` (which forwards to `isInvalidPreimageRootReason`) seem to have similar checks. They both check the `missing_validators` array.

Actually thinking about this, maybe we can just remove `preimage_root` from the `ConsensusData`? The root can be easily calculated after consensus is reached, when the block is about to be created and signed. So we wouldn't have to check its validity here, we would just have to compare the two arrays. What do you think, is it a good idea?

In fact I think we only need to verify the root when downloading blocks from other nodes.

linked0 replied:
> maybe we can just remove preimage_root from the ConsensusData?

I think we can do it because all the nodes have the same enrolled UTXOs (a.k.a. Validators) and we can calculate the root only with the indices. It's a good idea.

> In fact I think we only need to verify the root when downloading blocks from other nodes.

In fact, how to process the catch-up is a blocker for me now. There is some problem in checking the validity of the root in the middle of downloading blocks from other nodes. Suppose that the latest height is 1009 and all the current validators have enrolled at the height of 1008. When downloading the blocks between 0 and 1008, we can not check the validity of the root because there are no pre-images for un-enrolled validators.

AndrejMitrovic replied:
> there are no pre-images for un-enrolled validators.

That's why we should keep the last preimage and enrollment of every historic validator. It should be possible if we follow my new suggestion here: https://github.com/bpfkorea/agora/issues/1297#issuecomment-742203550

Then full nodes which want to validate really old blocks where those validators don't exist anymore would just use the `getPreimage` API to find those outdated validator's last preimages.

The problem right now is that we delete items from the `validator_set` as soon as a node gets un-enrolled. But we should probably keep that data forever so historic blocks can be validated. Of course, you would still need some information about which validators are **inactive**. Currently they're marked inactive by removing them from the validator set database, but maybe we should just add a status flag instead?

linked0 replied:
> Currently they're marked inactive by removing them from the validator set database, but maybe we should just add a status flag instead?

The state could solve my problem. But we should think about the situation that the data of `validator_set` will be ever-increasing, which might not be so big though.

AndrejMitrovic reviewed (changes requested):
Completed first round of review. Will wait for some comments.

AndrejMitrovic commented:
Great work so far @linked0!

linked0 committed:
ConsensusData nomination and validation

A `Validator` creates and nominates a `ConsensusData` struct with the
indices of the UTXOs of validators that are missing their preimages
(a.k.a. MPV). Whea a `Validator` receives a `ConsensusData` struct it
will try to verify that the `MPV` in the `ConsensusData` is the same
as the local MPV.

linked0 committed:
Add network test for `ConsensusData` validation

In fact, this test code might be better to be merged into the previous commit.
But, it is splitted due to so much code change in the previous commit.

linked0 commented:
Ready for review after applying the reviewers' suggestions, the flaky failures on the CI builds keep happening though.
I'll keep checking it.

AndrejMitrovic commented:
> the flaky failures on the CI builds keep happening though.

> /home/runner/work/agora/agora/submodules/ocean/src/ocean/text/convert/Formatter.d:466: error: undefined reference to '_D6geod247bitblob__T7BitBlobVmi512ZQp8toStringMxFNfMDFNfAxaZvZv'

It seems it's some kind of linker error.

linked0 commented:
> > the flaky failures on the CI builds keep happening though.
> 
> > /home/runner/work/agora/agora/submodules/ocean/src/ocean/text/convert/Formatter.d:466: error: undefined reference to '_D6geod247bitblob__T7BitBlobVmi512ZQp8toStringMxFNfMDFNfAxaZvZv'
> 
> It seems it's some kind of linker error.

Can I get any clue why the CI build on my PR keeps failing while the build on Chris's is going well now? The build on my local machine is going well.

linked0 commented:
I missed rebasing it. After rebasing it, I'm trying to ci build again.
I found with Chris's advice that it occurs only on the command of "dub build -c client --skip-registry=all --compiler=ldc2".

AndrejMitrovic commented:
Maybe you need to add more modules to https://github.com/bpfkorea/agora/blob/ea9f26c3d09527f79c16365633923b472393571e/dub.json#L96?

linked0 commented:
> Maybe you need to add more modules to
> 
> https://github.com/bpfkorea/agora/blob/ea9f26c3d09527f79c16365633923b472393571e/dub.json#L96
> 
> ?

will check it, thanks.

linked0 commented:
The ci builds are going well.
The position of the "version (unittest):" line in a network test file was wrong.

AndrejMitrovic commented on source/agora/consensus/protocol/Data.d:
```diff
@@ -13,6 +13,7 @@
 
 module agora.consensus.protocol.Data;
 
+import agora.common.Types;
```
Unnecessary import?

AndrejMitrovic commented on source/agora/consensus/SlashPolicy.d:
```diff
@@ -0,0 +1,289 @@
+
+        foreach (idx, utxo_key; keys)
+        {
+            if (!hasRevealedPreimage(height, utxo_key))
```
`this.hasRevealedPreimage`

AndrejMitrovic commented on source/agora/consensus/SlashPolicy.d:
```diff
@@ -0,0 +1,289 @@
+        Hash[] valid_keys;
+        foreach (key; keys)
+        {
+            if (hasRevealedPreimage(height, key))
```
`this.hasRevealedPreimage`

AndrejMitrovic commented on source/agora/consensus/SlashPolicy.d:
```diff
@@ -141,6 +142,11 @@ public class SlashPolicy
+        // NOTE: The preimage root of `Hash.init` value is currently
+        // checked in the `validateSlashingData` function of `Ledger`.
+        if (valid_keys.length == 0)
+            return Hash.init;
```
This should be documented in the `Returns` section.

AndrejMitrovic commented on source/agora/consensus/SlashPolicy.d:
```diff
@@ -169,6 +175,48 @@ public class SlashPolicy
+            assert(0);
+        }
+
+        Hash[] valid_keys;
```
This doesn't seem to be used for anything other than appending to it. I think it should be removed.

AndrejMitrovic commented on source/agora/consensus/SlashPolicy.d:
```diff
@@ -169,6 +175,48 @@ public class SlashPolicy
+            if (!hasRevealedPreimage(height, key))
+                local_missing_validators ~= cast(uint)idx;
+
+            if (missing_validators.find(idx).empty())
+                valid_keys ~= key;
```
Ditto, this should be removed.

AndrejMitrovic commented on source/agora/consensus/SlashPolicy.d:
```diff
@@ -169,6 +175,48 @@ public class SlashPolicy
+        uint[] local_missing_validators;
+        foreach (idx, key; keys)
+        {
+            if (!hasRevealedPreimage(height, key))
```
`this.hasRevealedPreimage`

AndrejMitrovic commented on source/agora/consensus/SlashPolicy.d:
```diff
@@ -169,6 +175,48 @@ public class SlashPolicy
+
+    ***************************************************************************/
+
+    public string isInvalidPreimageRootReason (Height height,
```
In the future it would be more efficient to construct the local array **once**, each time a new preimage is received. That way you can cache it and then the comparison is simple:

```
remote_missing == local_missing
```

linked0 replied:
> In the future it would be more efficient to construct the local array once, each time a new preimage is received. That way you can cache it and then the comparison is simple:
> 
> remote_missing == local_missing

Sure, I will. We had a similar requirement for the `preimage_root` which is not used though. We can apply the same concept for the variable for the `MPV`.

AndrejMitrovic reviewed (changes requested):
Just a few comments.

linked0 commented:
Ready for review.
- removed unnecessary import and variable
- added more explanation for the function description
- applied requests about the code

AndrejMitrovic commented on source/agora/consensus/protocol/Nominator.d:
```diff
@@ -544,6 +552,14 @@ extern(D):
                     fail_reason, data);
                 return ValidationLevel.kInvalidValue;
             }
+
```
You could do one more check here. If the array is bigger than the total number of active validators, then it's clearly invalid. So in that case you can return `kInvalidValue`.

linked0 replied:
> You could do one more check here. If the array is bigger than the total number of active validators, then it's clearly invalid. So in that case you can return kInvalidValue.

Ah, Right. It should return `kInvalidValue` in that case. I will fix it with the other requirements you suggested.
Thank you.

AndrejMitrovic commented:
There are some possible improvements, but I don't want to delay this any longer so I'm approving it.

AndrejMitrovic commented:
@linked0 I'll merge this so we can move forward with it!
